src/getDataFromTree.ts

```diff
@@ -1,275 +1,133 @@
-    rootNewContext,
-  );
+  public hasPromises() {
+    return this.queryPromises.size > 0;
+  }
 
-  return promises;
+  public consumeAndAwaitPromises() {
+    const promises: Promise<any>[] = [];
+    this.queryPromises.forEach((promise, queryInstance) => {
+      const { query, variables } = queryInstance.props;
+      // Make sure we never try to call fetchData for this query document and
+      // these variables again. Since the queryInstance objects change with
+      // every rendering, deduplicating them by query and variables is the
+      // best we can do. If a different Query component happens to have the
+      // same query document and variables, it will be immediately rendered
+      // by calling finish() in addQueryPromise, which could result in the
+      // rendering of an unwanted loading state, but that's not nearly as bad
+      // as getting stuck in an infinite rendering loop because we kept calling
+      // queryInstance.fetchData for the same Query component indefinitely.
+      this.queryGraveyard.add(query, JSON.stringify(variables));
+      promises.push(promise);
+    });
+    this.queryPromises.clear();
+    return Promise.all(promises);
+  }
 }
 
-function getDataAndErrorsFromTree(
-  rootElement: React.ReactNode,
-  rootContext: Object,
-  storeError: Function,
-  rootNewContext: Map<any, any> = new Map(),
-): Promise<any> {
-  const promises = getPromisesFromTree({ rootElement, rootContext, rootNewContext });
+class RenderPromisesProvider extends React.Component<{
+  renderPromises: RenderPromises;
+}> {
+  static childContextTypes = {
+    renderPromises: PropTypes.object,
+  };
 
-  if (!promises.length) {
-    return Promise.resolve();
+  getChildContext() {
+    return {
+      renderPromises: this.props.renderPromises,
+    };
   }
 
-  const mappedPromises = promises.map(({ promise, context, instance, newContext }) => {
-    return promise
-      .then(_ => getDataAndErrorsFromTree(instance.render(), context, storeError, newContext))
-      .catch(e => storeError(e));
-  });
-
-  return Promise.all(mappedPromises);
-}
-
-function processErrors(errors: any[]) {
-  switch (errors.length) {
-    case 0:
-      break;
-    case 1:
-      throw errors.pop();
-    default:
-      const wrapperError: any = new Error(
-        `${errors.length} errors were thrown when executing your fetchData functions.`,
-      );
-      wrapperError.queryErrors = errors;
-      throw wrapperError;
+  render() {
+    return this.props.children;
   }
 }
 
 export default function getDataFromTree(
   rootElement: React.ReactNode,
-  rootContext: any = {},
-): Promise<any> {
-  const errors: any[] = [];
-  const storeError = (error: any) => errors.push(error);
+  // The rendering function is configurable! We use renderToStaticMarkup as
+  // the default, because it's a little less expensive than renderToString,
+  // and legacy usage of getDataFromTree ignores the return value anyway.
+  renderFunction = renderToStaticMarkup,
```

Comment:
As a side note, it would be really awesome if dynamic `import()` was safe for libraries to use already, since it would enable us to call `import("react-dom/server")` here only if no `renderFunction` is provided, so `react-dom/server` and its dependencies would not be forced into the initial JS bundle. Since `getDataFromTree` is already async, a one-time dynamic `import()` would be pretty much invisible.

Reply:
##### Notes:
There is a breaking change of `getDataFromTree` in `react-apollo^2.3.0`. 
The second argument `rootContext` is replaced with `renderFunction`.

Reply:
Good point!

A few thoughts:
1. That's why it's a minor version bump.
2. I will have to think about whether `RenderPromisesProvider` can still provide a root context now that we don't have complete control over how contexts are managed.
3. Did you have any use cases for `rootContext`?

Reply:
Here is an example how I use `react-apollo v2.2.4` in next.js:
 
```js
await getDataFromTree(
  <ApolloProvider client={apollo}>
    <Component />
  </ApolloProvider>,
  { router }
)
```

ref: https://github.com/zeit/next.js/blob/canary/examples/with-apollo-and-redux/lib/withApollo.js#L42-L53

My workaround for `react-apollo v2.3.0`:
```diff
await getDataFromTree(
  <ApolloProvider client={apollo}>
+   <RouterProvider router={router}>
      <Component />
+   </RouterProvider>
  </ApolloProvider>,
- { router }
)
```

RouterProvider: https://github.com/zeit/next.js/issues/2908#issuecomment-333107793

Reply:
@evenchange4 Does this PR resolve your concerns? https://github.com/apollographql/react-apollo/pull/2586

Reply:
Great. Thank for your awesome work!